src/java.base/share/classes/java/util/Locale.java

```diff
@@ -980,33 +981,50 @@ static Locale getInstance(String language, String script, String country,
         return getInstance(baseloc, extensions);
     }
 
+
     static Locale getInstance(BaseLocale baseloc, LocaleExtensions extensions) {
         if (extensions == null) {
             Locale locale = CONSTANT_LOCALES.get(baseloc);
             if (locale != null) {
                 return locale;
             }
-            return LOCALE_CACHE.computeIfAbsent(baseloc, Locale::createLocale);
+            return LOCALE_CACHE.computeIfAbsent(baseloc, LOCALE_LOOKUP);
         } else {
             LocaleKey key = new LocaleKey(baseloc, extensions);
-            return LOCALE_CACHE.computeIfAbsent(key, Locale::createLocale);
+            return LOCALE_CACHE.computeIfAbsent(key, LOCALE_LOOKUP);
         }
     }
 
-    private static final ReferencedKeyMap<Object, Locale> LOCALE_CACHE = ReferencedKeyMap.create(true, ConcurrentHashMap::new);
-    private static Locale createLocale(Object key) {
-        if (key instanceof BaseLocale base) {
-            return new Locale(base, null);
+    private static final ReferencedKeyMap<Object, Locale> LOCALE_CACHE
+            = ReferencedKeyMap.create(true, ReferencedKeyMap.concurrentHashMapSupplier());
+
+    private static final LocaleKey LOCALE_LOOKUP = new LocaleKey();
```

Comment:
Reusing an existing class with a special instance for functional interface implementation is weird, I still wonder if we can do the local class + singleton (essentially same as non-capturing lambda) approach.